### DISCUSSION TITLE
map_overlap to slide over the image?

### DISCUSSION THREAD

**Discussion by Shaunakde** [dask/dask#7063] [Q&A]
I have a function that takes a chunk and computes a scalar, which I wrap into an array of shape (1,1,1). 

Using `map_blocks`, I am able to compute the result pretty easily. 

```
output = da.map_blocks(compute_fn, dcube.data, chunks=(1,1,1), dtype='float64').compute()
```
and the output has the shape `(1, 92, 144)`

Now I try to do the same thing with `map_overlap` and the `overlap` function. First, I do this:
```
overlap_dcube = da.overlap.overlap(dcube3.data, depth=({0:0, 1:9, 2:9}), boundary=0)
```

then call the `map_overlap` function:
```
da.map_overlap(cauchy_schwartz, overlap_dcube, depth=({0:1, 1:9, 2:9}), chunks=(1,1,1), dtype='float64').compute()
```
However, this results in the same as me applying `map_blocks` (only with different results because of the overlap step. 
What I actually want to achieve is to slide over the image with a stride of 1 (so that the output image is the same size as input).

However, I don't think I know to use the `map_overlap` or `rolling` function to achieve this correctly.  Any help would be greatly appreciated.

**Comment by jsignell** [comment]
When you say "stride of 1" do you mean 1 chunk or 1 data value? Right now it seems like you are getting a scalar per chunk. If you want to have the output be the same shape as the input then you need to adjust your function to return that. Maybe that looks like applying your current function several times within each map_overlap call.

**Comment by Shaunakde** [comment]
Dear @jsignell 

Apologies for not being very clear in my last message. Perhaps this diagram will help explain things a bit:
![image](https://user-images.githubusercontent.com/2545129/104476382-08aca980-5575-11eb-9ab1-dc7247b25dc4.png)

Basically, I have a dask array that has size (2,N,M) and a chunk size of (2,n,m). Each chunk produces a scalar - and I want to slide/stride by window, 1 pixel at a time. Your suggestion of nested `map_overlays` is a good one!

This is roughly how I have written it in scikit-learn for now:
```
A = view_as_windows(dcube3.data[0].compute(),(9,9))
B = view_as_windows(dcube3.data[1].compute(),(9,9))
C=np.zeros(img.shape)
for i in range(B.shape[0]):
    for j in range(B.shape[1]):
        C[i][j] = cauchy_schwartz_2(A[i][j], B[i][j])
```

Hope my intent is a little clearer. I think the linked issue I could find on this subject is this one: https://github.com/dask/dask/issues/4659

**Reply by jsignell** [reply]
Ah ok thanks for adding the visual - that really helped! I don't have any additional ideas right now, but I think what you have makes sense.
